### PR TITLE
Feat: Adding WIP Damage

### PR DESCRIPTION
**Pull request by Sparkymod**
- Add: a `Priest` character. (it covers almost all situation on skills)
- Add: `SkillCast` class for the skill that is being use in the moment.
- Update: SkillUse, SkillHandler and SkillDamage to use SkillCast.
- Add: `DamageHandler` class container of the damage
- Add: default items for Priest character
- Update: `stats.hp.max = stats.hp.total` for mobs. In order to make calculations, it has to substract/add from `Max`. Then `Total` is used for the hp bar.
- Fix: `SkillJobSub` for those job that has sub skill in `Xml/table/job` but that id isn't an actual existing skill on `Xml/skill`
- Fix: Mobs now disappear from field when they die

### CODE REVIEW

File: MapleServer2/Packets/SkillDamagePacket.cs [hunk 1]
```diff
@@ -46,5 +53,6 @@ public static Packet ApplyHeal(IFieldObject<Player> player, int statusUid, int s
             pWriter.WriteByte(1);
             return pWriter;
         }
+
```
Comment (angelotadeucci) [inline MapleServer2/Packets/SkillDamagePacket.cs:56]: new line

File: MapleServer2/Packets/StatPacket.cs [hunk 2]
```diff
@@ -26,9 +26,12 @@ public static Packet UpdateMobStats(IFieldObject<Mob> mob)
             pWriter.WriteByte(4); // value
             // Stats 
             // Damage should be update through this packet
-            for (int i = 0; i < 3; i++)
+            pWriter.WriteLong(mob.Value.Stats.Hp.Total);
+            pWriter.WriteLong(mob.Value.Stats.Hp.Min);
+            pWriter.WriteLong(mob.Value.Stats.Hp.Max -= (long) skillCast.GetDamage());
+            if (mob.Value.Stats.Hp.Max <= 0)
```
Comment (angelotadeucci) [inline MapleServer2/Packets/StatPacket.cs:32]: I'm sure we don't want logic inside the packets classes.

File: MapleServer2/Types/Mob.cs [hunk 3]
```diff
@@ -6,12 +6,24 @@ namespace MapleServer2.Types
     public class Mob : NpcMetadata
     {
         public short ZRotation; // In degrees * 10
+        private bool IsDead;
```
Comment (angelotadeucci) [inline MapleServer2/Types/Mob.cs:9]: Why not use `private bool IsDead {get; set;}` ?

File: MapleServer2/Types/Player.cs [hunk 4]
```diff
@@ -239,7 +237,65 @@ public static Player NewCharacter(byte gender, Job job, string name, SkinColor s
...
+                },
+                IsVIP = false,
+            };
+            player.Equips.Add(ItemSlot.RH, Item.DefaultScepter(player));
+            player.Equips.Add(ItemSlot.LH, Item.DefaultCodex(player));
+            return player;
+        }
+
         public void Warp(MapPlayerSpawn spawn, int mapId)
+
```
Comment (angelotadeucci) [inline MapleServer2/Types/Player.cs:298]: new line

File: MapleServer2/Packets/StatPacket.cs [hunk 5]
```diff
@@ -26,9 +26,12 @@ public static Packet UpdateMobStats(IFieldObject<Mob> mob)
             pWriter.WriteByte(4); // value
             // Stats 
             // Damage should be update through this packet
-            for (int i = 0; i < 3; i++)
+            pWriter.WriteLong(mob.Value.Stats.Hp.Total);
+            pWriter.WriteLong(mob.Value.Stats.Hp.Min);
+            pWriter.WriteLong(mob.Value.Stats.Hp.Max -= (long) skillCast.GetDamage());
```
Comment (AlanMorel) [inline MapleServer2/Packets/StatPacket.cs:31]: can we move this and the below code out the packet method?

File: MapleServer2/Types/SkillCast.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        public short GetSkillLevel()
+        {
+            return SkillLevel;
+        }
+
+        public int GetUnkValue()
+        {
+            return UnkValue;
+        }
```
Comment (angelotadeucci) [inline MapleServer2/Types/SkillCast.cs:34]: Also here, why not use `{ get; }` ?

File: MapleServer2/Types/SkillCast.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            Random rnd = new Random();
+            double roll = rnd.NextDouble();
+            if (roll > 0.5)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
```
Comment (angelotadeucci) [inline MapleServer2/Types/SkillCast.cs:60]: `return roll > 0.5 ? true : false`

File: MapleServer2/Types/SkillCast.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            // TODO: Critic Damage calculation
+            if (RollCrit() == true)
+            {
+                return GetDamageRate() + GetDamageRate();
+            }
+            else
+            {
+                return GetDamageRate();
+            }
```
Comment (angelotadeucci) [inline MapleServer2/Types/SkillCast.cs:73]: `return RollCrit() ? GetDamageRate() * 2 : GetDamageRate()`

File: MapleServer2/Types/SkillCast.cs (hunk 7 above)
Comment (Sparkymod) [reply MapleServer2/Types/SkillCast.cs:60]: sure, forgot that exist.

File: MapleServer2/Types/Player.cs [hunk 9]
```diff
@@ -239,7 +237,65 @@ public static Player NewCharacter(byte gender, Job job, string name, SkinColor s
             };
         }
 
+        public static Player Priest(long accountId, long characterId, string name = "Priest")
+        {
+            Job job = Job.Priest; // Priest
```
Comment (AlanMorel) [inline MapleServer2/Types/Player.cs:242]: Not sure we need this comment

File: MapleServer2/Types/SkillCast.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using MapleServer2.Data.Static;
+
+namespace MapleServer2.Types
+{
+    public class SkillCast
+    {
+        #region Declaration
```
Comment (AlanMorel) [inline MapleServer2/Types/SkillCast.cs:8]: Can we remove regions?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/SkillCast.cs (hunk 6 above)
Comment (Sparkymod) [reply MapleServer2/Types/SkillCast.cs:34]: sure!, it was a way to understand what i was doing, i can simplify.

**Review by angelotadeucci — APPROVED** [review]

File: MapleServer2/Types/SkillCast.cs [hunk 11]
```diff
@@ -50,34 +30,14 @@ public bool RollCrit()
...
             double roll = rnd.NextDouble();
-            if (roll > 0.5)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
+            return roll > 0.5 ? true : false;
```
Comment (AlanMorel) [inline MapleServer2/Types/SkillCast.cs:33]: Don't need the true/false part here

File: MapleServer2/Packets/StatPacket.cs [hunk 12]
```diff
@@ -17,27 +17,23 @@ public static Packet SetStats(IFieldObject<Player> player)
...
             pWriter.WriteByte(4); // value
-            // Stats 
-            // Damage should be update through this packet
-            for (int i = 0; i < 3; i++)
-            {
-                pWriter.WriteLong(mob.Value.Stats.Hp[i]);
-            }
+            pWriter.WriteLong(mob.Value.Stats.Hp.Total);
+            pWriter.WriteLong(mob.Value.Stats.Hp.Min);
+            pWriter.WriteLong(mob.Value.UpdateStats(mob.Value, skillCast.GetDamage()));
```
Comment (AlanMorel) [inline MapleServer2/Packets/StatPacket.cs:29]: Again, why are we changing mob stats in the middle of the packet? Can this be done outside the packet method? Also, why do we need to pass in `mob.Value` when we already have access to it

File: MapleServer2/Types/DamageHandler.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+        public static DamageHandler CalculateSkillDamage(SkillCast skillCast)
+        {
+            double damage;
+            damage = skillCast.GetDamageRate();
+            DamageHandler toReturn = HandleCrit(skillCast, damage);
+            return toReturn;
+        }
+
+
```
Comment (AlanMorel) [inline MapleServer2/Types/DamageHandler.cs:45]: Blank lines

File: MapleServer2/Types/Mob.cs [hunk 14]
```diff
@@ -1,17 +1,27 @@
...
 using MapleServer2.Data.Static;
 
 namespace MapleServer2.Types
 {
     public class Mob : NpcMetadata
     {
         public short ZRotation; // In degrees * 10
+        public bool IsDead { get; private set; }
+
+        public long UpdateStats(Mob mob, double damage)
```
Comment (AlanMorel) [inline MapleServer2/Types/Mob.cs:12]: Why's this method above the constructor?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/Mob.cs [hunk 15]
```diff
@@ -25,5 +17,12 @@ public Mob(int id)
             Stats = mob.Stats;
             Friendly = mob.Friendly;
         }
+
+        public void UpdateStats(double damage)
+        {
+            Mob mob = this;
```
Comment (AlanMorel) [inline MapleServer2/Types/Mob.cs:23]: You don't need this line, you should be able to just access `Stats` and `IsDead` directly here

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/Mob.cs (hunk 15 above)
Comment (Sparkymod) [reply MapleServer2/Types/Mob.cs:23]: sure! yeah, i though i need the same value where is used but is in the same class doesn't need it.

**Review by angelotadeucci — APPROVED** [review]

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**